File: stroom-query-api/src/main/java/stroom/query/api/ExpressionBuilder.java
/*
 * Copyright 2017 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package stroom.query.api;


import stroom.query.api.ExpressionOperator.Op;
import stroom.query.api.ExpressionTerm.Condition;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class ExpressionBuilder {
    private final Boolean enabled;
    private final Op op;

    private final List<Object> children = new ArrayList<>();

    /**
     * By default an expression builder will create enabled operators and will 'AND' all added terms.
     */
    public ExpressionBuilder() {
        this(null, Op.AND);
    }

    public ExpressionBuilder(final Op op) {
        this(null, op);
    }

    public ExpressionBuilder(final Boolean enabled, final Op op) {
        this.enabled = enabled;
        this.op = op;
    }

    public ExpressionBuilder addTerm(final String field, final Condition condition, final String value) {
        return addTerm(null, field, condition, value, null);
    }

    public ExpressionBuilder addTerm(final Boolean enabled, final String field, final Condition condition, final String value) {
        return addTerm(enabled, field, condition, value, null);
    }

    public ExpressionBuilder addTerm(final String field, final Condition condition, final DocRef dictionary) {
        return addTerm(null, field, condition, null, dictionary);
    }

    public ExpressionBuilder addTerm(final Boolean enabled, final String field, final Condition condition, final DocRef dictionary) {
        return addTerm(enabled, field, condition, null, dictionary);
    }

    public ExpressionBuilder addTerm(final Boolean enabled, final String field, final Condition condition, final String value, final DocRef dictionary) {
        children.add(new ExpressionTerm(enabled, field, condition, value, dictionary));
        return this;
    }

    public ExpressionBuilder addTerm(final ExpressionTerm term) {
        if (term != null) {
            addTerm(term.getEnabled(), term.getField(), term.getCondition(), term.getValue(), term.getDictionary());
        }

        return this;
    }

    public ExpressionBuilder addOperator(final Op op) {
        return addOperator(null, op);
    }

    public ExpressionBuilder addOperator(final Boolean enabled, final Op op) {
        final ExpressionBuilder builder = new ExpressionBuilder(enabled, op);
        children.add(builder);
        return builder;
    }

    public ExpressionBuilder addOperator(final ExpressionOperator operator) {
        if (operator != null) {
            final ExpressionBuilder builder = addOperator(operator.getEnabled(), operator.getOp());

            if (operator.getChildren() != null) {
                for (final ExpressionItem child : operator.getChildren()) {
                    if (child instanceof ExpressionOperator) {
                        builder.addOperator((ExpressionOperator) child);
                    } else if (child instanceof ExpressionTerm) {
                        builder.addTerm((ExpressionTerm) child);
                    }
                }
            }

            return builder;
        }

        return null;
    }

    public ExpressionOperator build() {
        if (children.size() == 0) {
            return new ExpressionOperator(enabled, op, Arrays.asList());
        }

        final List<ExpressionItem> list = new ArrayList<>(children.size());
        for (final Object object : children) {
            if (object instanceof ExpressionBuilder) {
                list.add(((ExpressionBuilder) object).build());
            } else if (object instanceof ExpressionTerm) {
                list.add((ExpressionTerm) object);
            }
        }

        return new ExpressionOperator(enabled, op, list);
    }
}

File: stroom-ui/src/components/ExpressionBuilder/ExpressionBuilder.js
/*
 * Copyright 2018 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';

import { compose, withState, branch, renderComponent, withProps, lifecycle } from 'recompose';
import { connect } from 'react-redux';
import { Loader, Checkbox } from 'semantic-ui-react';

import ExpressionOperator from './ExpressionOperator';
import ROExpressionOperator from './ROExpressionOperator';
import { LineContainer } from 'components/LineTo';
import DeleteExpressionItem from './DeleteExpressionItem';

import lineElementCreators from './expressionLineCreators';

const withSetEditableByUser = withState('inEditMode', 'setEditableByUser', false);

const ROExpressionBuilder = ({ expressionId, expression }) => (
  <LineContainer
    lineContextId={`expression-lines-${expressionId}`}
    lineElementCreators={lineElementCreators}
  >
    <ROExpressionOperator expressionId={expressionId} isEnabled operator={expression} />
  </LineContainer>
);

ROExpressionBuilder.propTypes = {
  expressionId: PropTypes.string.isRequired,
  expression: PropTypes.object.isRequired,
};

const enhance = compose(
  connect(
    ({ expressionBuilder }, { expressionId }) => ({
      expressionState: expressionBuilder[expressionId],
    }),
    {
      // actions
    },
  ),
  withSetEditableByUser,
  lifecycle({
    componentDidMount() {
      this.props.setEditableByUser(this.props.editMode);
    },
  }),
  branch(
    ({ expressionState }) => !expressionState,
    renderComponent(() => <Loader active>Loading Expression State</Loader>),
  ),
  withProps(({ showModeToggle, dataSource }) => ({
    showModeToggle: showModeToggle && !!dataSource,
  })),
);

const ExpressionBuilder = ({
  expressionId,
  dataSource,
  expressionState: { expression },
  showModeToggle,
  inEditMode,
  setEditableByUser,
}) => (
  <LineContainer
    className="Expression-editor__graph"
    lineContextId={`expression-lines-${expressionId}`}
    lineElementCreators={lineElementCreators}
  >
    <DeleteExpressionItem expressionId={expressionId} />
    {showModeToggle ? (
      <Checkbox
        label="Edit Mode"
        toggle
        checked={inEditMode}
        onChange={() => setEditableByUser(!inEditMode)}
      />
    ) : (
      undefined
    )}
    {inEditMode ? (
      <ExpressionOperator
        dataSource={dataSource}
        expressionId={expressionId}
        isRoot
        isEnabled
        operator={expression}
      />
    ) : (
      <ROExpressionOperator expressionId={expressionId} isEnabled operator={expression} />
    )}
  </LineContainer>
);

const EnhancedExpressionBuilder = enhance(ExpressionBuilder);

EnhancedExpressionBuilder.propTypes = {
  dataSource: PropTypes.object,
  expressionId: PropTypes.string.isRequired,
  showModeToggle: PropTypes.bool.isRequired,
  editMode: PropTypes.bool,
};

EnhancedExpressionBuilder.defaultProps = {
  showModeToggle: false,
};

export default EnhancedExpressionBuilder;
